core/blockchain/validate.go

```diff
@@ -1102,23 +1127,52 @@ func (b *BlockChain) CheckTransactionInputs(tx *types.Tx, utxoView *UtxoViewpoin
 	// Calculate the total output amount for this transaction.  It is safe
 	// to ignore overflow and out of range errors here because those error
 	// conditions would have already been caught by checkTransactionSanity.
-	var totalAtomOut int64
+	totalAtomOut := make(map[types.CoinID]int64)
 	for _, txOut := range tx.Transaction().TxOut {
-		totalAtomOut += int64(txOut.Amount) //TODO, remove type conversion
+		// Ensure the coinId is known
+		err := types.CheckCoinID(txOut.Amount.Id)
+		if err != nil {
+			return 0, err
+		}
+		totalAtomOut[txOut.Amount.Id] += txOut.Amount.Value
 	}
 
+	// Ensure no unbalanced/unknowned coin type from input/output
+	if len(totalAtomIn) != len(totalAtomOut) ||
+		len(totalAtomIn) > len(types.CoinIDList) ||
+		len(totalAtomOut) > len(types.CoinIDList) {
+		return 0, fmt.Errorf("transaction contains unknown or unbalanced coin types")
+	}
 	// Ensure the transaction does not spend more than its inputs.
-	if totalAtomIn < totalAtomOut {
-		str := fmt.Sprintf("total value of all transaction inputs for "+
-			"transaction %v is %v which is less than the amount "+
-			"spent of %v", txHash, totalAtomIn, totalAtomOut)
-		return 0, ruleError(ErrSpendTooHigh, str)
+	for _, coinId := range types.CoinIDList {
+		if totalAtomIn[coinId] < totalAtomOut[coinId] {
+			str := fmt.Sprintf("total %s value of all transaction inputs for "+
+				"transaction %v is %v which is less than the amount "+
+				"spent of %v", coinId.Name(), txHash, totalAtomIn[coinId], totalAtomOut[coinId])
+			return 0, ruleError(ErrSpendTooHigh, str)
+		}
 	}
 
 	// NOTE: bitcoind checks if the transaction fees are < 0 here, but that
 	// is an impossible condition because of the check above that ensures
 	// the inputs are >= the outputs.
-	txFeeInAtom := totalAtomIn - totalAtomOut
+
+	// check if txfee in meer
+	txFeeInAtom := totalAtomIn[types.MEERID] - totalAtomOut[types.MEERID]
```

Comment:
what if the transaction doesn't include meer, is it possible?

Reply:
it possible, it does not force tx to include a meer input. it will first check the txfee(meer), if meer is zero. try another coin type. it allows the tx pay the fee by using any coin type at the current stage. the txfee rule may and should be changed in the future.

Reply:
So if totalAtomIn doesn't include meer, would totalAtomIn[types.MEERID] throw out of index exception?

Reply:
totalAtomIn map will result a zero value if not found. will not result an out of index.
```golang
m := make(map[string]float64)

m["pi"] = 3.14             // Add a new key-value pair

v := m["pi"]               // Get value: v == 3.1416
v = m["pie"]               // Not found: v == 0 (zero value)
```

Reply:
Indeed, my mistake, I just made a test as well, basically I will use v, ok = m["pie"]  to see if pie exists